src/services/goToDefinition.ts

```diff
@@ -68,11 +68,10 @@ namespace ts.GoToDefinition {
         //      bar<Test>(({pr/*goto*/op1})=>{});
         if (isPropertyName(node) && isBindingElement(parent) && isObjectBindingPattern(parent.parent) &&
             (node === (parent.propertyName || parent.name))) {
+            const name = getNameFromPropertyName(node);
             const type = typeChecker.getTypeAtLocation(parent.parent);
-            const propSymbols = getPropertySymbolsFromType(type, node);
-            if (propSymbols) {
-                return flatMap(propSymbols, propSymbol => getDefinitionFromSymbol(typeChecker, propSymbol, node));
-            }
+            const propSymbols = name === undefined ? emptyArray : mapDefined(type.isUnion() ? type.types : [type], t => t.getProperty(name));
+            return flatMap(propSymbols, propSymbol => getDefinitionFromSymbol(typeChecker, propSymbol, node));

```

Comment:
You dont need to create propSymbols as well right? You can just move this lambda above ?